### ISSUE TITLE
NeedBraces: lambda with no braces not reported when contents span multiple lines

### ISSUE THREAD

**Issue by rnveach** [checkstyle/checkstyle#3837]
````
$ cat TestClass.java
public class TestClass {
    void method() {
        final Set<FullIdent> imports = new HashSet<>();

        imports.stream().filter(full -> test(1, 2,
            3, 4));
    }

    boolean test(int p1, int p2, int p3, int p4) { return true; }
}

$ cat TestConfig.xml
<?xml version="1.0"?>
<!DOCTYPE module PUBLIC
          "-//Puppy Crawl//DTD Check Configuration 1.3//EN"
          "http://www.puppycrawl.com/dtds/configuration_1_3.dtd">

<module name="Checker">
    <property name="charset" value="UTF-8"/>

    <module name="TreeWalker">
    <module name="NeedBraces">
      <property name="tokens" value="LAMBDA"/>
      <property name="allowSingleLineStatement" value="true"/>
    </module>
    </module>
</module>

$ java -jar checkstyle-7.5.1-all.jar -c TestConfig.xml TestClass.java
Starting audit...
Audit done.
````

`allowSingleLineStatement` is turned on, but this lambda isn't a single line as the method call for `test` spans 2 lines and is inside the lambda.
I was expecting a violation to be reported on line 5.

Example is taken from checkstyle at https://github.com/checkstyle/checkstyle/blob/b4a337a599e853cdc44b19a76da28fcfab2956cd/src/main/java/com/puppycrawl/tools/checkstyle/checks/imports/RedundantImportCheck.java#L126-L129. We are using this same check and configuration.

**Comment by romani** [comment]
pre-assigned to @johnyleebrown .

**Comment by johnyleebrown** [comment]
I'm on it.

**Comment by johnyleebrown** [comment]
@rnveach I've been trying to find a solution, so i started by editing method `isSingleLineLambda(DetailAST lambda)`:
<img width="1115" alt="screen shot 2017-03-20 at 01 01 18" src="https://cloud.githubusercontent.com/assets/10441153/24085214/d25fe1fc-0d08-11e7-9cfe-e1517feb8840.png">
But when i'm trying to test, this error comes up, can you tell me what does it relate to?Thanks.
<img width="773" alt="screen shot 2017-03-20 at 00 58 01" src="https://cloud.githubusercontent.com/assets/10441153/24085190/5633c652-0d08-11e7-80e2-769d054d0c99.png">
Without line 420 everything tests fine.

**Comment by johnyleebrown** [comment]
@rnveach maybe I should create a test explicitly for multiple lines lambda, woth the code I added?

I'll create a pr later today.

**Comment by rnveach** [comment]
> maybe I should create a test explicitly for multiple lines lambda

Yes, we expect good cases (no violation) and bad cases (violation). Example from issue should be in the test cases and any others that you can generate that are pertinent.

**Comment by johnyleebrown** [comment]
@rnveach ,got it. Thanks.

**Comment by romani** [comment]
issue is assigned to @sharang108 .

**Comment by sharang108** [comment]
@romani I am on it!

**Comment by rnveach** [comment]
@romani How do we find the end of the line for the lambda?
Based on the AST tree, it is embedded in the method call.
Print:
````
    |      |      |--ELIST -> ELIST [7:37]
    |      |      |  `--LAMBDA -> -> [7:37]
    |      |      |      |--IDENT -> full [7:32]
    |      |      |      `--EXPR -> EXPR [7:44]
    |      |      |          `--METHOD_CALL -> ( [7:44]
    |      |      |              |--IDENT -> test [7:40]
    |      |      |              |--ELIST -> ELIST [7:45]
    |      |      |              |  |--EXPR -> EXPR [7:45]
    |      |      |              |  |  `--NUM_INT -> 1 [7:45]
    |      |      |              |  |--COMMA -> , [7:46]
    |      |      |              |  |--EXPR -> EXPR [7:48]
    |      |      |              |  |  `--NUM_INT -> 2 [7:48]
    |      |      |              |  |--COMMA -> , [7:49]
    |      |      |              |  |--EXPR -> EXPR [8:12]
    |      |      |              |  |  `--NUM_INT -> 3 [8:12]
    |      |      |              |  |--COMMA -> , [8:13]
    |      |      |              |  `--EXPR -> EXPR [8:15]
    |      |      |              |      `--NUM_INT -> 4 [8:15]
    |      |      |              `--RPAREN -> ) [8:16]
    |      |      `--RPAREN -> ) [8:17]
````

**Comment by romani** [comment]
@rnveach ,
```
    |      |      |  `--LAMBDA -> -> [7:37]
.......
    |      |      |              `--RPAREN -> ) [8:16]
```

start at line 7 finished at 8, look simple.
am I missing smth ?

**Comment by rnveach** [comment]
At the time I assumed ``--RPAREN -> ) [8:17]` was the end since it wraps the lambda.
So we just keep getting the last child of `LAMBDA` until the end?

**Comment by rnveach** [comment]
Fix was merged
